rule prinseq_plus_plus_fas2fq:
    input:
        "reads/{prefix}.fas",
    output:
        good="results/{prefix}.fq",
        bad="results/{prefix}.bad.fq",
    log:
        "logs/fas2fq/{prefix}.log",
    params:
        extra="-min_len 2",
    threads: 2
    wrapper:
        "master/bio/prinseq-plus-plus"


rule prinseq_plus_plus_fas2fqgz:
    input:
        "reads/{prefix}.fas",
    output:
        good="results/{prefix}.fq.gz",
        bad="results/{prefix}.bad.fq.gz",
    log:
        "logs/fas2fqgz/{prefix}.log",
    params:
        extra="-min_len 2",
    threads: 2
    wrapper:
        "master/bio/prinseq-plus-plus"


rule prinseq_plus_plus_fqgz2fas:
    input:
        "reads/{prefix}.fastq.gz",
    output:
        good="results/{prefix}.fasta",
        bad="results/{prefix}.bad.fasta",
    log:
        "logs/fqgz2fas/{prefix}.log",
    params:
        extra="-min_len 2",
    threads: 2
    wrapper:
        "master/bio/prinseq-plus-plus"


rule prinseq_plus_plus_fq2fasgz:
    input:
        "reads/{prefix}.fastq",
    output:
        good="results/{prefix}.fas.gz",
        bad="results/{prefix}.bad.fas.gz",
    log:
        "logs/fq2fasgz/{prefix}.log",
    params:
        extra="-min_len 2",
    threads: 2
    wrapper:
        "master/bio/prinseq-plus-plus"


rule prinseq_plus_plus_fqpe:
    input:
        "reads/{prefix}.1.fastq.gz",
        "reads/{prefix}.2.fastq.gz",
    output:
        good="results/{prefix}.R1.fq.gz",
        good2="results/{prefix}.R2.fq.gz",
        single="results/{prefix}.single.R1.fq.gz",
        single2="results/{prefix}.single.R2.fq.gz",
        bad="results/{prefix}.bad.R1.fq.gz",
        bad2="results/{prefix}.bad.R2.fq.gz",
    log:
        "logs/fqpe/{prefix}.log",
    params:
        extra="-min_len 2",
    threads: 2
    wrapper:
        "master/bio/prinseq-plus-plus"
