rule star_pe_multi:
    input:
        # use a list for multiple fastq files for one sample
        # usually technical replicates across lanes/flowcells
        fq1=["reads/{sample}_R1.1.fastq", "reads/{sample}_R1.2.fastq"],
        # paired end reads needs to be ordered so each item in the two lists match
        fq2=["reads/{sample}_R2.1.fastq", "reads/{sample}_R2.2.fastq"],  #optional
    output:
        # see STAR manual for additional output files
        sam="star/pe/{sample}/Aligned.out.sam",
        log="star/pe/{sample}/Log.out",
    log:
        "logs/star/pe/{sample}.log",
    params:
        # path to STAR reference genome index
        idx="index",
        # optional parameters
        extra="",
    threads: 8
    wrapper:
        "master/bio/star/align"


rule star_se:
    input:
        fq1="reads/{sample}_R1.1.fastq",
    output:
        # see STAR manual for additional output files
        sam="star/se/{sample}/Aligned.out.sam",
        log="star/se/{sample}/Log.out",
    log:
        "logs/star/se/{sample}.log",
    params:
        # path to STAR reference genome index
        idx="index",
        # optional parameters
        extra="",
    threads: 8
    wrapper:
        "master/bio/star/align"
