rule mehari_annotate_seqvars_variants_MT:
    input:
        calls="{prefix}.vcf",  # .vcf, .vcf.gz or .bcf
        reference_fasta="resources/MT.fasta",  # has to be uncompressed and must have corresponding index genome.fasta.fai
        transcript_db="resources/MT-ND2-GRCh38-ensembl-0.10.3.bin.zst",  # transcript database for SO term / consequence annotation
        # clinvar_db="resources/clinvar.bin.zst",  # clinvar database for clinvar VCV annotation
        # frequency_db="resources/frequencies.bin.zst"  # frequencies/gnomad database for frequency annotation
    output:
        calls="{prefix}.annotated.bcf",  # .vcf, .vcf.gz or .bcf
    params:
        extra="",  # extra options include: --report-most-severe-consequence-by {gene, transcript, allele}, --pick-transcript {mane-select, mane-plus-clinical, length, ensembl-canonical, ref-seq-select, gencode-primary, basic}, --pick-transcript-mode {first, all}, --keep-intergenic, --discard-utr-splice-variants
    log:
        "logs/mehari/mehari_annotate_variants.{prefix}.log",
    wrapper:
        "master/bio/mehari/annotate-seqvars"
