rule get_variation:
    output:
        vcf="refs/variation.vcf.gz"
        # Optional: add fai to get VCF with annotated contig lengths (as required by GATK)
        # and properly sorted VCFs.
        # fai="refs/genome.fasta.fai"
    params:
        species="saccharomyces_cerevisiae",
        release="98", # releases <98 are unsupported
        build="R64-1-1",
        type="all" # one of "all", "somatic", "structural_variation"
    log:
        "logs/get_variation.log"
    cache: True  # save space and time with between workflow caching (see docs)
    wrapper:
        "master/bio/reference/ensembl-variation"
