
rule all:  # [hide]
    input:  # [hide]
        "results/multiqc/broadPeak/multiqc_report.html",  # [hide]


rule get_genome:  # [hide]
    output:  # [hide]
        "data/genome.fasta",  # [hide]
    conda:
        "envs/curl.yaml"
    log:
        "logs/get-genome.log"
    shell:  # [hide]
        "curl -L https://raw.githubusercontent.com/nf-core/test-datasets/atacseq/reference/genome.fa > {output} 2> {log}"  # [hide]


rule get_annotation:  # [hide]
    output:  # [hide]
        "data/genome.gtf",  # [hide]
    conda:
        "envs/curl.yaml"
    log:
        "logs/get-annotation.log"
    shell:  # [hide]
        "curl -L https://raw.githubusercontent.com/nf-core/test-datasets/atacseq/reference/genes.gtf > {output} 2> {log}"  # [hide]


rule get_design:  # [hide]
    output:  # [hide]
        "design.csv",  # [hide]
    conda:
        "envs/curl.yaml"
    log:
        "logs/get-design.log"
    shell:  # [hide]
        "curl -L https://raw.githubusercontent.com/nf-core/test-datasets/chipseq/design.csv > {output} 2> {log}"  # [hide]


rule chipseq_pipeline:
    input:
        input="design.csv",
        fasta="data/genome.fasta",
        gtf="data/genome.gtf",
        # any --<argname> pipeline file arguments can be given here as <argname>=<path>
    output:
        "results/multiqc/broadPeak/multiqc_report.html",
    params:
        pipeline="nf-core/chipseq",
        revision="1.2.1",
        profile=["test", "docker"],
        # any --<argname> pipeline arguments can be given here as <argname>=<value>
    handover: True
    wrapper:
        "master/utils/nextflow"
