rule vg_autoindex_giraffe:
    input:
        ref="{genome}.fasta",
        # Optional vcf file containing haplotypes for graph creation
        #vcf="",
    output:
        multiext("resources/{genome}", ".dist", ".min", ".giraffe.gbz"),
    log:
        "logs/autoindex/{genome}.log",
    params:
        extra="",
    threads: 8
    wrapper:
        "master/bio/vg/autoindex"


rule vg_autoindex_map:
    input:
        ref="{genome}.fasta",
        # Optional vcf file containing haplotypes for graph creation
        #vcf="",
    output:
        multiext("resources/{genome}", ".xg", ".gcsa", ".gcsa.lcp"),
    log:
        "logs/autoindex/{genome}.log",
    params:
        extra="",
    threads: 8
    wrapper:
        "master/bio/vg/autoindex"
