rule quast:
    input:
        fasta="genome.fasta",
        ref="genome.fasta",
        #gff="annotations.gff",
        #pe1="reads_R1.fastq",
        #pe2="reads_R2.fastq",
        #pe12="reads.fastq",
        #mp1="matereads_R1.fastq",
        #mp2="matereads_R2.fastq",
        #mp12="matereads.fastq",
        #single="single.fastq",
        #pacbio="pacbio.fas",
        #nanopore="nanopore.fastq",
        #ref_bam="ref.bam",
        #ref_sam="ref.sam",
        #bam=["s1.bam","s2.bam"],
        #sam=["s1.sam","s2.sam"],
        #sv_bedpe="sv.bed",
    output:
        report_html="{sample}/report.html",
        report_tex="{sample}/report.tex",
        report_txt="{sample}/report.txt",
        report_pdf="{sample}/report.pdf",
        report_tsv="{sample}/report.tsv",
        treport_tex="{sample}/treport.tex",
        treport_txt="{sample}/treport.txt",
        treport_tsv="{sample}/treport.tsv",
        stats_cum="{sample}/stats/cumulative.pdf",
        stats_gc_plot="{sample}/stats/gc.pdf",
        stats_gc_icarus="{sample}/stats/gc.icarus.txt",
        stats_gc_genome="{sample}/stats/gc_genome.pdf",
        stats_ngx="{sample}/stats/NGx.pdf",
        stats_nx="{sample}/stats/Nx.pdf",
        contigs="{sample}/contigs.all_alignments.tsv",
        contigs_mis="{sample}/contigs.mis_contigs.info",
        icarus="{sample}/icarus.html",
        icarus_viewer="{sample}/icarus_viewer.html",
    log:
        "logs/{sample}.log",
    params:
        extra="--min-contig 5 --min-identity 95.0",
    wrapper:
        "master/bio/quast"
