rule hla_candidates:
    input:
        genome="genome_hla.fasta",
        alleles="hla_alleles.txt",
        xml="hla.xml"
    output:
        "results/hla_candidates/"  # Directory output
    log:
        "logs/orthanq_candidates/candidates_hla.log",
    params:
        subcommand="hla",
        output_bcf=True  # or False
    wrapper:
        "master/bio/orthanq/hla/candidates"

rule virus_candidates:
    input:
        genome="genome_virus.fasta",
        lineages="lineages.txt"
    output:
        vcf="results/virus_candidates.vcf"
    log:
        "logs/orthanq_candidates/candidates_virus.log",
    params:
        subcommand="virus"
    wrapper:
        "master/bio/orthanq/hla/candidates"

