rule samtools_sort:
    input:
        "{path}/{sample}.bam",
    output:
        temp("{path}/{sample}.sorted.bam"),
    params:
        extra="-m 4G",
        tmp_dir="/tmp/",
    log:
        "{path}_{sample}.log",
    # Samtools takes additional threads through its option -@
    threads: 8  # This value - 1 will be sent to -@.
    wrapper:
        "master/bio/samtools/sort"


rule calc_consensus_reads:
    input:
        "mapped/{sample}.marked.sorted.bam",
    output:
        # non-overlapping consensus read pairs will be written into consensus_r1 and consensus_r2 
        consensus_r1=temp("results/consensus/{sample}.1.fq"),
        consensus_r2=temp("results/consensus/{sample}.2.fq"),
        # consensus reads from single end records or overlapping read pairs will be merged into a single end record
        consensus_se=temp("results/consensus/{sample}.se.fq"),
        # skipped reads (soft-clipped or unpropper mapped reads) will be skipped and unmarked
        skipped=temp("results/consensus/{sample}.skipped.bam"),
    params:
        extra="",
    log:
        "logs/consensus/{sample}.log",
    wrapper:
        "master/bio/rbt/collapse_reads_to_fragments-bam"


rule map_consensus_reads:
    input:
        reads=lambda wc: expand(
            "results/consensus/{sample}.{read}.fq",
            sample=wc.sample,
            read="se" if wc.read_type == "se" else (1, 2),
        ),
        idx=multiext("resources/genome.fa", ".amb", ".ann", ".bwt", ".pac", ".sa"),
    output:
        temp("results/consensus/{sample}.{read_type}.mapped.bam"),
    params:
        index=lambda w, input: os.path.splitext(input.idx[0])[0],
        sort="samtools",
        sort_order="coordinate",
    wildcard_constraints:
        read_type="pe|se",
    log:
        "logs/bwa_mem/{sample}.{read_type}.consensus.log",
    threads: 8
    wrapper:
        "master/bio/bwa/mem"


rule mark_skipped_duplicates:
    input:
        bams=["results/consensus/{sample}.sorted.bam"],
    output:
        bam=temp("results/consensus/{sample}.marked.bam"),
        metrics="results/consensus/{sample}.metrics.txt",
    log:
        "logs/picard/marked/{sample}.log",
    params:
        extra="--VALIDATION_STRINGENCY LENIENT --TAG_DUPLICATE_SET_MEMBERS true",
    # optional specification of memory usage of the JVM that snakemake will respect with global
    # resource restrictions (https://snakemake.readthedocs.io/en/latest/snakefiles/rules.html#resources)
    # and which can be used to request RAM during cluster job submission as `{resources.mem_mb}`:
    # https://snakemake.readthedocs.io/en/latest/executing/cluster.html#job-properties
    resources:
        mem_mb=1024,
    wrapper:
        "master/bio/picard/markduplicates"


rule merge_consensus_reads:
    input:
        "results/consensus/{sample}.skipped.marked.bam",
        "results/consensus/{sample}.se.mapped.bam",
        "results/consensus/{sample}.pe.mapped.bam",
    output:
        "results/consensus/{sample}.merged.bam",
    log:
        "logs/samtools_merge/{sample}.log",
    threads: 8
    wrapper:
        "master/bio/samtools/merge"
