rule vg_autoindex:  # [hide]
    input:  # [hide]
        ref="{genome}.fasta",  # [hide]
    output:  # [hide]
        multiext("{genome}", ".dist", ".shortread.withzip.min", ".giraffe.gbz"),  # [hide]
    log:  # [hide]
        "logs/vg_autoindex/{genome}.log",  # [hide]
    params:  # [hide]
        extra="",  # [hide]
    threads: 8  # [hide]
    wrapper:  # [hide]
        "master/bio/vg/autoindex"  # [hide]


rule vg_giraffe_map:
    input:
        reads=["reads/{sample}.1.fastq", "reads/{sample}.2.fastq"],
        graph="resources/a.gbz",
        #kmers="resources/{sample}.kff", # optional: kmer counts
        #hapl="resources/{genome}.hapl", # optional: haplotype index
    output:
        bam="mapped/{sample}.bam",
        indexes=multiext(
            "resources/{sample}",
            ".dist",
            ".shortread.withzip.min",
            ".shortread.zipcodes",
        ),
    log:
        "logs/vg_giraffe/{sample}.log",
    params:
        extra="",
        sorting="none",  # Can be 'none', 'samtools' or 'picard'.
        sort_order="queryname",  # Can be 'queryname' or 'coordinate'.
        sort_extra="",  # Extra args for samtools/picard.
    threads: 8
    wrapper:
        "master/bio/vg/giraffe"
