rule test_agat_config:
    output:
        "out/agat_config.yaml",
    threads: 1
    log:
        "logs/test_agat_config.log",
    params:
        command="config",
    wrapper:
        "master/bio/agat"


rule test_agat_levels:
    output:
        "out/agat_levels.yaml",
    log:
        "logs/test_agat_levels.log",
    params:
        command="levels",
    wrapper:
        "master/bio/agat"


rule test_agat_convert_bed2gff:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        bed="input/test.bed",
    output:
        # The key here will be used as argument name in the final command line
        output="out/test_agat_convert_bed2gff.gff",
    log:
        "logs/test_agat_convert_bed2gff.log",
    params:
        command="agat_convert_bed2gff.pl",
        extra="",  # Anything besides IO
    wrapper:
        "master/bio/agat"


rule test_agat_convert_embl2gff:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        embl="input/agat_convert_embl2gff_1.embl",
    output:
        # The key here will be used as argument name in the final command line
        output="out/test_agat_convert_embl2gff.gff",
    log:
        "logs/test_agat_convert_embl2gff.log",
    params:
        command="agat_convert_embl2gff.pl",
        extra="",  # Anything besides IO
    wrapper:
        "master/bio/agat"


rule test_agat_convert_genscan2gff:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        genscan="input/test.genscan",
    output:
        # The key here will be used as argument name in the final command line
        output="out/test_agat_convert_genscan2gff.gff",
    log:
        "logs/test_agat_convert_genscan2gff.log",
    params:
        command="agat_convert_genscan2gff.pl",
        extra="",  # Anything besides IO
    wrapper:
        "master/bio/agat"


rule test_agat_convert_mfannot2gff:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        mfannot="input/test.mfannot",
    output:
        # The key here will be used as argument name in the final command line
        output="out/test_agat_convert_mfannot2gff.gff",
    log:
        "logs/agat_convert_mfannot2gff.log",
    params:
        # No optional non-file parameters with this subcommand.
        command="agat_convert_mfannot2gff.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_convert_minimap2_bam2gff_bam:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        i="input/test_minimap2.bam",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_convert_minimap2_bam2gff_bam.gff",
    log:
        "logs/agat_convert_minimap2_bam2gff_bam.log",
    params:
        # All non-file parameters are auto detected.
        command="agat_convert_minimap2_bam2gff.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_convert_minimap2_bam2gff_sam:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        i="input/test_minimap2.sam",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_convert_minimap2_bam2gff_sam.gff",
    log:
        "logs/agat_convert_minimap2_bam2gff_sam.log",
    params:
        # All non-file parameters are auto detected.
        command="agat_convert_minimap2_bam2gff.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_convert_sp_gff2bed:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_convert_sp_gff2bed.bed",
    log:
        "logs/agat_convert_sp_gff2bed.log",
    params:
        command="agat_convert_sp_gff2bed.pl",
        extra="",  # Non file arguments (`--nc` or `--sub`).
    wrapper:
        "master/bio/agat"


rule test_agat_convert_sp_gff2gtf:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_convert_sp_gff2gtf.gtf",
    log:
        "logs/agat_convert_sp_gff2gtf.log",
    params:
        command="agat_convert_sp_gff2gtf.pl",
        extra="",  # Non file arguments (`--gtf_version`).
    wrapper:
        "master/bio/agat"


rule test_agat_convert_sp_gff2tsv:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_convert_sp_gff2tsv.tsv",
    log:
        "logs/agat_convert_sp_gff2tsv.log",
    params:
        # No non-file parameters for this command
        command="agat_convert_sp_gff2tsv.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_convert_sp_gff2zff:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # The key here will be used as argument name in the final command line
        ann="out/test_agat_convert_sp_gff2zff.ann",
        dna="out/test_agat_convert_sp_gff2zff.dna",
    log:
        "logs/agat_convert_sp_gff2zff.log",
    params:
        # No non-file parameters for this command
        command="agat_convert_sp_gff2zff.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_convert_sp_gxf2gxf:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_convert_sp_gxf2gxf.gff",
    log:
        "logs/agat_convert_sp_gxf2gfx.log",
    params:
        command="agat_convert_sp_gxf2gxf.pl",
        extra="",  # Non file arguments (`--verbose`)
    wrapper:
        "master/bio/agat"


rule test_agat_sp_prokka_infer_name_from_attributes:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_sp_prokka_infer_name_from_attributes.gff",
    log:
        "logs/agat_sp_prokka_infer_name_from_attributes.log",
    params:
        command="agat_sp_prokka_infer_name_from_attributes.pl",
        extra="",  # Non file arguments (`--force`).
    wrapper:
        "master/bio/agat"


rule test_agat_sp_add_intergenic_regions:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_sp_add_intergenic_regions.gff",
    log:
        "logs/test_agat_sp_add_intergenic_regions.log",
    params:
        command="agat_sp_add_intergenic_regions.pl",
        extra="",  # Non file arguments (`--verbose`).
    wrapper:
        "master/bio/agat"


rule test_agat_sp_add_introns:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_sp_add_introns.gff",
    log:
        "logs/test_agat_sp_add_introns.log",
    params:
        # No non-file arguments.
        command="agat_sp_add_introns.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_add_splice_sites:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_sp_add_splice_sites.gff",
    log:
        "logs/test_agat_sp_add_splice_sites.log",
    params:
        # No non-file arguments.
        command="agat_sp_add_splice_sites.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_add_start_and_stop:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_sp_add_start_and_stop.gff",
    log:
        "logs/test_agat_sp_add_start_and_stop.log",
    params:
        extra="",  # Non-file arguments (e.g. `--codon`, `--extend`, `--verbose`, or `--na`)
        command="agat_sp_add_start_and_stop.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_alignment_output_style:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_sp_alignment_output_style.gff",
    log:
        "logs/test_agat_sp_alignment_output_style.log",
    params:
        extra="",  # Non-file arguments (e.g. `--ct`, or `--verbose`)
        command="agat_sp_alignment_output_style.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_clipN_seqExtremities_and_fixCoordinates:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # The key here will be used as argument name in the final command line
        og="out/test_agat_sp_clipN_seqExtremities_and_fixCoordinates.gff",
        of="out/test_agat_sp_clipN_seqExtremities_and_fixCoordinates.fasta",
    log:
        "logs/test_agat_sp_clipN_seqExtremities_and_fixCoordinates.log",
    params:
        extra="",  # Non-file arguments (e.g. `--of`, or `--og`)
        command="agat_sp_clipN_seqExtremities_and_fixCoordinates.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_compare_two_annotations:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff1="input/annotation.gff",
        gff2="input/prokka_fragmented_genes.gff",
    output:
        # The key here will be used as argument name in the final command line
        # This subcommand returns a file per case, which names are build based
        # on the comparison result as well as a report.
        out=directory("out/test_agat_sp_compare_two_annotations"),
    log:
        "logs/test_agat_sp_compare_two_annotations.log",
    params:
        extra="",  # Non-file arguments (`--verbose`)
        command="agat_sp_compare_two_annotations.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_complement_annotations:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        ref="input/annotation.gff",
        add="input/prokka_fragmented_genes.gff",
    output:
        # The key here will be used as argument name in the final command line
        out="out/test_agat_sp_complement_annotations.gff",
    log:
        "logs/test_agat_sp_complement_annotations.log",
    params:
        extra="",  # Non-file arguments (`--size_min`)
        command="agat_sp_complement_annotations.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_ensembl_output_style:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        o="out/test_agat_sp_ensembl_output_style.gff",
    log:
        "logs/test_agat_sp_ensembl_output_style.log",
    params:
        extra="",  # Non-file arguments (`--ct`, or `--verbose`)
        command="agat_sp_ensembl_output_style.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_extract_attributes:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        Parent="out/test_agat_sp_extract_attributes_Parent.txt",
        ID="out/test_agat_sp_extract_attributes_ID.txt",
    log:
        "logs/test_agat_sp_extract_attributes.log",
    params:
        # Non-file arguments besides `--att` (e.g. `-p`, `--merge`, or `-d`)
        extra="-d",
        command="agat_sp_extract_attributes.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_extract_sequences:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # The key here will be used as argument name in the final command line
        o="out/test_agat_sp_extract_sequences.fasta",
    log:
        "logs/test_agat_sp_extract_attributes.log",
    params:
        # Non-file arguments (e.g. `--asc`, `--cdna`, `--cfs`, ...)
        extra="-p -t cds",
        command="agat_sp_extract_sequences.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_filter_by_ORF_size:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # Matched = ORF size satisfying `--test` and `--size` criteria
        matched="out/test_agat_sp_filter_by_ORF_size_matched.gff",
        # Unmatched = ORF size *NOT* satisfying `--test` and `--size` criteria
        unmatched="out/test_agat_sp_filter_by_ORF_size_unmatched.gff",
    log:
        "logs/test_agat_sp_filter_by_ORF_size.log",
    params:
        # Non-file arguments (e.g. `--size`, `--test`, or `--verbose`)
        extra="",
        command="agat_sp_filter_by_ORF_size.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_filter_by_locus_distance:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        output="out/test_agat_sp_filter_by_locus_distance.gff",
    log:
        "logs/test_agat_sp_filter_by_locus_distance.log",
    params:
        # Non-file arguments (e.g. `-d`, or `--add_flag`)
        extra="",
        command="agat_sp_filter_by_locus_distance.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_filter_feature_by_attribute_presence:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        o="out/test_agat_sp_filter_feature_by_attribute_presence.gff",
    log:
        "logs/test_agat_sp_filter_feature_by_attribute_presence.log",
    params:
        # Non-file arguments (`--type`, `--flip` or `--att`)
        extra="--type gene --att Parent",
        command="agat_sp_filter_feature_by_attribute_presence.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_filter_feature_by_attribute_value:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        o="out/test_agat_sp_filter_feature_by_attribute_value.gff",
    log:
        "logs/test_agat_sp_filter_feature_by_attribute_value.log",
    params:
        # Non-file arguments (e.g. `--type`, `--value`, `--value_insensitive`...)
        extra="--type exon --attribute constitutive --value 1",
        command="agat_sp_filter_feature_by_attribute_value.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_filter_feature_from_keep_list:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        keep_list="input/keep_list.txt",
    output:
        # The key here will be used as argument name in the final command line
        o="out/test_agat_sp_filter_feature_from_keep_list.gff",
    log:
        "logs/test_agat_sp_filter_feature_from_keep_list.log",
    params:
        # Non-file arguments (`--type`, or `--verbose`)
        extra="--verbose 4",
        command="agat_sp_filter_feature_from_keep_list.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_filter_feature_from_kill_list:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        kill_list="input/kill_list.txt",
    output:
        # The key here will be used as argument name in the final command line
        o="out/test_agat_sp_filter_feature_from_kill_list.gff",
    log:
        "logs/test_agat_sp_filter_feature_from_kill_list.log",
    params:
        # Non-file arguments (`--type`, or `--verbose`)
        extra="--verbose 4",
        command="agat_sp_filter_feature_from_kill_list.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_filter_gene_by_intron_numbers:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        o="out/test_agat_sp_filter_gene_by_intron_numbers.gff",
    log:
        "logs/test_agat_sp_filter_gene_by_intron_numbers.log",
    params:
        # Non-file arguments (`--number`, `--verbose`, or `--test`)
        extra="--test '>=' --nb 3",
        command="agat_sp_filter_gene_by_intron_numbers.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_filter_gene_by_length:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line
        o="out/test_agat_sp_filter_gene_by_length.gff",
    log:
        "logs/test_agat_sp_filter_gene_by_length.log",
    params:
        # Non-file arguments (`--size`, `--verbose`, or `--test`)
        extra="--test '>=' --size 3",
        command="agat_sp_filter_gene_by_length.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_filter_incomplete_gene_coding_models:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # The key here will be used as argument name in the final command line
        output="out/test_agat_sp_filter_incomplete_gene_coding_models.gff",
    log:
        "logs/test_agat_sp_filter_incomplete_gene_coding_models.log",
    params:
        # Non-file arguments (e.g. `--codon`, `--add_flag`, `--skip_start_check`, ...)
        extra="--verbose 4",
        command="agat_sp_filter_incomplete_gene_coding_models.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_filter_record_by_coordinates:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        coordinates="input/coordinates.tsv",
    output:
        # The key here will be used as argument name in the final command line.
        # The directory contains a file per interval (if any record exists within),
        # a GFF with records belonging to none of the intervald given in coordinates,
        # and a text report.
        o=directory("out/test_agat_sp_filter_record_by_coordinates"),
    log:
        "logs/test_agat_sp_filter_record_by_coordinates.log",
    params:
        # Non-file arguments (`--exclude` or `--verbose`)
        extra="--verbose 4",
        command="agat_sp_filter_record_by_coordinates.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_fix_cds_phases:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # The key here will be used as argument name in the final command line.
        o="out/test_agat_sp_fix_cds_phases.gff",
    log:
        "logs/test_agat_sp_fix_cds_phases.log",
    params:
        # Non-file arguments (`--verbose`)
        extra="",
        command="agat_sp_fix_cds_phases.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_fix_features_locations_duplicated:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line.
        output="out/test_agat_sp_fix_features_locations_duplicated.gff",
    log:
        "logs/test_agat_sp_fix_features_locations_duplicated.log",
    params:
        # Non-file arguments (`--verbose`, or `--model`)
        extra="",
        command="agat_sp_fix_features_locations_duplicated.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_fix_fusion:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # Optional file contains all features in the GFF
        all="out/test_agat_sp_fix_fusion_all.gff",
        # Optional file only contains features that were modified by agat
        only_modified="out/test_agat_sp_fix_fusion_modified.gff",
        # Optional file only contains features that were *NOT* modified by agat
        intact="out/test_agat_sp_fix_fusion_intact.gff",
        # An optional report over the number of modified/intact features
        report="out/test_agat_sp_fix_fusion_report.txt",
    log:
        "logs/test_agat_sp_fix_fusion.log",
    params:
        # Non-file arguments (`--codon`, `--threshold`, `--stranded`, or `--verbose`)
        extra="",
        command="agat_sp_fix_fusion.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_fix_longest_ORF:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # Optional file contains all features in the GFF
        all="out/test_agat_sp_fix_longest_ORF_all.gff",
        # Optional file only contains features that were modified by agat
        only_modified="out/test_agat_sp_fix_longest_ORF_modified.gff",
        # Optional file only contains features that were *NOT* modified by agat
        intact="out/test_agat_sp_fix_longest_ORF_intact.gff",
        # An optional report over the number of modified/intact features
        report="out/test_agat_sp_fix_longest_ORF_report.txt",
    log:
        "logs/test_agat_sp_fix_longest_ORF.log",
    params:
        # Non-file arguments (`--codon`, `--split`, `--stranded`, or `--verbose`)
        extra="",
        command="agat_sp_fix_longest_ORF.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_fix_overlaping_genes:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line.
        output="out/test_agat_sp_fix_overlaping_genes.gff",
    log:
        "logs/test_agat_sp_fix_overlaping_genes.log",
    params:
        # Non-file arguments (`--merge` or `--verbose`)
        extra="",
        command="agat_sp_fix_overlaping_genes.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_fix_small_exon_from_extremities:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # The key here will be used as argument name in the final command line.
        output="out/test_agat_sp_fix_small_exon_from_extremities.gff",
    log:
        "logs/test_agat_sp_fix_small_exon_from_extremities.log",
    params:
        # Non-file arguments (`--merge`, `--size`, `--codon`, or `--verbose`)
        extra="",
        command="agat_sp_fix_small_exon_from_extremities.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_flag_premature_stop_codons:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # The key here will be used as argument name in the final command line.
        out="out/test_agat_sp_flag_premature_stop_codons.gff",
    log:
        "logs/test_agat_sp_flag_premature_stop_codons.log",
    params:
        extra="",  # Non-file argument: `--codon`
        command="agat_sp_flag_premature_stop_codons.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_flag_short_introns:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line.
        out="out/test_agat_sp_flag_short_introns.gff",
    log:
        "logs/test_agat_sp_flag_short_introns.log",
    params:
        # Non-file parameters (`--intron_size` or `--verbose`)
        extra="",
        command="agat_sp_flag_short_introns.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_functional_statistics:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line.
        # This directory contains sub-directory for each type of features available
        # in the GFF/GTF file, as long as a report
        o=directory("out/test_agat_sp_functional_statistics"),
    log:
        "logs/test_agat_sp_functional_statistics.log",
    params:
        # Non-file parameters (`--gs`)
        extra="",
        command="agat_sp_functional_statistics.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_keep_longest_isoform:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line.
        o="out/test_agat_sp_keep_longest_isoform.gff",
    log:
        "logs/test_agat_sp_keep_longest_isoform.log",
    params:
        # No non-file parameters allowed
        command="agat_sp_keep_longest_isoform.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_kraken_assess_liftover:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gtf="input/test_kraken.gtf",
    output:
        # The key here will be used as argument name in the final command line.
        out="out/test_agat_sp_kraken_assess_liftover.gff",
    log:
        "logs/test_agat_sp_kraken_assess_liftover.log",
    params:
        # Non-file parameters (`--threshold` or `--verbose`)
        extra="",
        command="agat_sp_kraken_assess_liftover.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_list_short_introns:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/agat_sq_add_attributes_from_tsv.gff",
    output:
        # The key here will be used as argument name in the final command line.
        out="out/test_agat_sp_list_short_introns.gff",
    log:
        "logs/test_agat_sp_list_short_introns.log",
    params:
        # Non-file parameter: `--size`
        extra="",
        command="agat_sp_list_short_introns.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_manage_IDs:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/agat_sq_add_attributes_from_tsv.gff",
    output:
        # The key here will be used as argument name in the final command line.
        out="out/test_agat_sp_manage_IDs.gff",
    log:
        "logs/test_agat_sp_manage_IDs.log",
    params:
        # Non-file parameters (e.g. `--gap`, `--ensembl`, `--prefix`, ...)
        extra="",
        command="agat_sp_manage_IDs.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_manage_UTRs:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # Optional path to PDF histogram of 5'UTR with at least <n> exons
        five_prime_utr_overORequal="out/test_agat_sp_manage_UTRs_five_prime_utr_overORequal.pdf",
        # Optional path to PDF histogram of 5'UTR with less than <n> exons
        five_prime_utr_under="out/test_agat_sp_manage_UTRs_five_prime_utr_under.pdf",
        # Optional path to PDF histogram of 3'UTR with at least <n> exons
        three_prime_utr_overORequal="out/test_agat_sp_manage_UTRs_three_prime_utr_overORequal.pdf",
        # Optional path to PDF histogram of 3'UTR with at least <n> exons
        three_prime_utr_under="out/test_agat_sp_manage_UTRs_three_prime_utr_under.pdf",
        # Optional path to PDF histogram of both 3/5'UTR with at least <n> exons
        # both_utr_overORequal="out/test_agat_sp_manage_UTRs_both_utr_overORequal.pdf",
        # Optional path to PDF histogram of both 3/5'UTR with at least <n> exons
        # both_utr_under="out/test_agat_sp_manage_UTRs_both_utr_under.pdf",
        # Optional path to analysis report
        report="out/test_agat_sp_manage_UTRs_report.txt",
    log:
        "logs/test_agat_sp_manage_UTRs.log",
    params:
        # Non-file parameters (e.g. `--number`, `--three`, `--five`, or `--both`)
        # The parameter `--plot` is automatically infered
        extra="--three --five -n 6",
        command="agat_sp_manage_UTRs.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_manage_attributes:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/agat_sq_add_attributes_from_tsv.gff",
    output:
        # The key here will be used as argument name in the final command line.
        out="out/test_agat_sp_manage_attributes.gff",
    log:
        "logs/test_agat_sp_manage_attributes.log",
    params:
        # Non-file parameters (`--type`, `--tag`, `--add`, `--cp`, or `--overwrite`)
        extra="--att biotype",
        command="agat_sp_manage_attributes.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_manage_functional_annotation:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/agat_sp_manage_functional_annotation/02413F.gff",
        # Optional path to blast, interpro, and reference fasta
        blast="input/agat_sp_manage_functional_annotation/02413F_blast.out",
        db="input/agat_sp_manage_functional_annotation/uniprot_sprot_test.fasta",
        interpro="input/agat_sp_manage_functional_annotation/02413F_interpro.tsv",
    output:
        # The key here will be used as argument name in the final command line.
        gff="out/test_agat_sp_manage_functional_annotation.gff",
        # Global text report
        report="out/test_agat_sp_manage_functional_annotation.report.txt",
        # Warnings and non-blocking format errors
        error="out/test_agat_sp_manage_functional_annotation.error.txt",
        # Duplicates blast sequences, requires blast and db
        duplicates="out/test_agat_sp_manage_functional_annotation.duplicates.tsv",
        # Conserved protein domains database terms, requires all optional input
        cdd="out/test_agat_sp_manage_functional_annotation.cdd.tsv",
        # Gene onthology database terms, requires all optional input
        go="out/test_agat_sp_manage_functional_annotation.go.tsv",
        # Interpro domain database terms, requires all optional input
        interpro="out/test_agat_sp_manage_functional_annotation.interpro.tsv",
        # MobiDB Lite database terms, requires all optional input
        mobidb="out/test_agat_sp_manage_functional_annotation.mobidblite.tsv",
        # Panther database terms, requires all optional input
        panther="out/test_agat_sp_manage_functional_annotation.panther.tsv",
        # Superfamily database terms, requires all optional input
        superfamily="out/test_agat_sp_manage_functional_annotation.superfamily.tsv",
    log:
        "logs/test_agat_sp_manage_functional_annotation.log",
    params:
        # Non-file parameters (e.g. `--blast_evalue`, `--pe`, `--clean_ontology`, ...)
        # The parameter `--plot` is automatically infered.
        extra="",
        command="agat_sp_manage_functional_annotation.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_manage_introns:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # Optional path to intron report
        report="out/test_agat_sp_manage_introns_report.txt",
        # Optional path to cds histogram (triggers the `--plot` argument)
        cds_pdf="out/test_agat_sp_manage_introns_cds.pdf",
        # Optional path to exon histogram (triggers the `--plot` argument)
        exon_pdf="out/test_agat_sp_manage_introns_exons.pdf",
        # Optional path to 3'UTR histogram (triggers the `--plot` argument)
        # three_prime_utr_pdf="out/test_agat_sp_manage_introns_exons_3p.pdf",
        # Optional path to 5'UTR histogram (triggers the `--plot` argument)
        five_prime_utr_pdf="out/test_agat_sp_manage_introns_exons_5p.pdf",
    log:
        "logs/test_agat_sp_manage_introns.log",
    params:
        # Non-file parameters (`--p`, `--window`)
        # The parameter `--plot` is automatically infered.
        extra="",
        command="agat_sp_manage_introns.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_merge_annotations:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff=["input/prokka_fragmented_genes.gff", "input/annotation.gff"],
    output:
        # The key here will be used as argument name in the final command line.
        out="out/test_agat_sp_merge_annotations.gff",
    log:
        "logs/test_agat_sp_merge_annotations.log",
    params:
        # No non-file parameters
        command="agat_sp_merge_annotations.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_move_attributes_within_records:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/agat_sq_add_attributes_from_tsv.gff",
    output:
        # The key here will be used as argument name in the final command line.
        out="out/test_agat_sp_move_attributes_within_records.gff",
    log:
        "logs/test_agat_sp_move_attributes_within_records.log",
    params:
        # Non-file parameters (`--fp`, `--fc`, `--attribute`, or `--verbose`)
        extra="--feature_copy mRNA  --feature_paste CDS",
        command="agat_sp_move_attributes_within_records.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_prokka_fix_fragmented_gene_annotations:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/prokka_fragmented_genes.gff",
        fasta="input/prokka_cav_10DC88.fa",
        db="input/prokka_bacteria_sprot.fa",
    output:
        # The key here will be used as argument name in the final command line.
        # The output directory contains a `report.txt` and additional files
        # depending on the content of input file(s).
        out=directory("out/test_agat_sp_prokka_fix_fragmented_gene_annotations"),
    log:
        "logs/test_agat_sp_prokka_fix_fragmented_gene_annotations.log",
    params:
        # Non-file parameters:
        # `--frags`, `--pseudo`, `--hmap_size`, `--ct`, `--skip_hmap`, or `--verbose`
        extra="--skip_hamap",  # To speed up tests
        command="agat_sp_prokka_fix_fragmented_gene_annotations.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_sensitivity_specificity:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff1="input/prokka_fragmented_genes.gff",
        gff2="input/annotation.gff",
    output:
        # The key here will be used as argument name in the final command line.
        output="out/test_agat_sp_sensitivity_specificity.txt",
    log:
        "logs/test_agat_sp_sensitivity_specificity.log",
    params:
        # Non-file parameters (`--verbose`)
        extra="",
        command="agat_sp_sensitivity_specificity.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_separate_by_record_type:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/agat_sq_add_attributes_from_tsv.gff",
    output:
        # The key here will be used as argument name in the final command line.
        # The output directory contains up to one file per record type among the
        # huge list of records available in `agat levels`.
        o=directory("out/test_agat_sp_separate_by_record_type"),
    log:
        "logs/test_agat_sp_separate_by_record_type.log",
    params:
        # No non-file parameters
        command="agat_sp_separate_by_record_type.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sp_statistics:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/agat_sq_add_attributes_from_tsv.gff",
    output:
        # Optional Path to text report
        report="out/test_agat_sp_statistics.txt",
        # Optional path to yaml report (triggers the `--yaml` argument)
        yaml="out/test_agat_sp_statistics.yaml",
        # Optional path to directory containing multiple plots, corresponding
        # to each level present in the input GFF file and described by the
        # `agat levels` command. Triggers `-p` argument.
        plot=directory("out/test_agat_sp_statistics_plots"),
    log:
        "logs/test_agat_sp_statistics.log",
    params:
        # Non-file parameters: `--gs`, or `--verbose`)
        # All other non-file parameters are automatically infered
        extra="--gs 432709230",
        command="agat_sp_statistics.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_add_attributes_from_tsv:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/agat_sq_add_attributes_from_tsv.gff",
        tsv="input/agat_sq_add_attributes_from_tsv.tsv",
    output:
        # Optional Path to text report
        o="out/test_agat_sq_add_attributes_from_tsv.gff",
    log:
        "logs/test_agat_sq_add_attributes_from_tsv.log",
    params:
        # Non-file parameters: `--verbose`
        # All other parameters are automatically included
        extra="",
        command="agat_sq_add_attributes_from_tsv.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_add_hash_tag:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/agat_sq_add_attributes_from_tsv.gff",
    output:
        # Optional Path to text report
        o="out/test_agat_sq_add_hash_tag.gff",
    log:
        "logs/test_agat_sq_add_hash_tag.log",
    params:
        extra="--interval 1",
        command="agat_sq_add_hash_tag.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_add_locus_tag:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/agat_sq_add_attributes_from_tsv.gff",
    output:
        # Optional Path to text report
        o="out/test_agat_sq_add_locus_tag.gff",
    log:
        "logs/test_agat_sq_add_locus_tag.log",
    params:
        # Non-file parameters, eg: `--type`, `--lo`, `--li`, `--of` or `--quiet`
        extra="",
        command="agat_sq_add_locus_tag.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_filter_feature_from_fasta:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
        fasta="input/sequence.fa",
    output:
        # Optional Path to text report
        o="out/test_agat_sq_filter_feature_from_fasta.gff",
    log:
        "logs/test_agat_sq_filter_feature_from_fasta.log",
    params:
        # Only verbosity parameter is expected, other ones are automatically filled
        extra="--verbose 4",
        command="agat_sq_filter_feature_from_fasta.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_list_attributes:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # Optional Path to attributes
        output="out/test_agat_sq_list_attributes.txt",
    log:
        "logs/test_agat_sq_list_attributes.log",
    params:
        # Only `-p`, `-t` or `-l`, other options are automatically infered
        extra="-p level2,cds,exon",
        command="agat_sq_list_attributes.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_manage_IDs:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # Optional Path to ids
        o="out/test_agat_sq_manage_IDs.txt",
    log:
        "logs/test_agat_sq_manage_IDs.log",
    params:
        # Only `--verbose`, all other parameters are infered
        extra="--verbose 4",
        command="agat_sq_manage_IDs.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_manage_attributes:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/annotation.gff",
    output:
        # Optional Path to GFF
        output="out/test_agat_sq_manage_attributes.gff",
    log:
        "logs/test_agat_sq_manage_attributes.log",
    params:
        # One of `--type`, `--tag` `--add`, `--vp`,
        # `--overwrite`, `--value` or `--strategy`
        extra="--tag locus_tag,product,name/NewName --type level2,cds,exon",
        command="agat_sq_manage_attributes.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_mask:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/prokka_fragmented_genes.gff",
        fasta="input/prokka_cav_10DC88.fa",
    output:
        # Path to masked GFF/GTF
        o="out/test_agat_sq_mask.gff",
    log:
        "logs/test_agat_sq_mask.log",
    params:
        # Only `--sm` or `--hm` (mutually exclusive), all other parameters are infered.
        extra="--hm",
        command="agat_sq_mask.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_remove_redundant_entries:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/prokka_fragmented_genes.gff",
    output:
        # Path to GFF entries
        o="out/test_agat_sq_remove_redundant_entries.gff",
    log:
        "logs/test_agat_sq_remove_redundant_entries.log",
    params:
        # All parameters are filled automatically
        command="agat_sq_remove_redundant_entries.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_repeats_analyzer:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/prokka_fragmented_genes.gff",
        # Optional path to fasta sequence, see params.extra
        genome="input/prokka_cav_10DC88.fa",
    output:
        # Path to repeats
        o="out/test_agat_sq_repeats_analyzer.gff",
    log:
        "logs/test_agat_sq_repeats_analyzer.log",
    params:
        # If no fasta file is provided in `input.genome`, then provide the genome size
        # through optional parameter below:
        # extra="--genome 125",
        command="agat_sq_repeats_analyzer.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_reverse_complement:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/prokka_fragmented_genes.gff",
        fasta="input/prokka_cav_10DC88.fa",
    output:
        # Path to revert complement
        o="out/test_agat_sq_reverse_complement.gff",
    log:
        "logs/test_agat_sq_reverse_complement.log",
    params:
        # Only verbosity parameter is allowed, other ones are automatically filled
        extra="--verbose 4",
        command="agat_sq_reverse_complement.pl",
    wrapper:
        "master/bio/agat"


rule test_agat_sq_rfam_analyzer:
    input:
        # The key here will be used as argument name in the final command line
        # config="", # Optional path to configuration file
        gff="input/prokka_fragmented_genes.gff",
        genome="input/prokka_cav_10DC88.fa",
    output:
        # Optional Path to ids
        o="out/test_agat_sq_rfam_analyzer.tsv",
    log:
        "logs/test_agat_sq_rfam_analyzer.log",
    params:
        # If no fasta file is provided in `input.genome`, then provide the genome size
        # through optional parameter below:
        # extra="--genome 125",
        command="agat_sq_rfam_analyzer.pl",
    wrapper:
        "master/bio/agat"
